src/nats/connection.cr

```diff
@@ -465,20 +473,28 @@ module NATS
       close
     end
 
-    @flush_counter = 0
-
     private def flush_outbound
       @out.synchronize do
         @socket.flush
+        @waiting_count.set 0
       end
     end
 
     private def outbound
       until closed?
-        fs = @flush.receive
-        break if fs.nil?
+        {
+          5.microseconds,
+          10.microseconds,
+          50.microseconds,
+          100.microseconds,
+          500.microseconds,
+          1.millisecond,
+          5.milliseconds,
+        }.each do |duration|
+          sleep duration
+          break if @waiting_count.get > 0
```

Comment:
This is not doing what we want. We want to us no cpu when nothing pending. When we get woke from a signal we should delay before sending as you do above. But we should break out before then end of all the timeouts if pre_waiting == waiting, meaning no more activity.

So in the case of a single message we should try to optimize for low latency. For a spike in publish calls we will keep seeing more waiting and moving onto the next duration.

Reply:
When there's nothing pending, though, that doesn't use zero CPU. In [this diff](https://github.com/jgaskins/nats.cr/compare/crystal-0.32-compatibility...jgaskins:crystal-0.32-compatibility-publish-backoff), I think I got it working [exactly like the Rust driver](https://github.com/nats-io/nats.rs/blob/99a5176ba218798eca5adadb185004ba45c0e0fc/src/lib.rs#L327-L355), and it uses quite a bit of CPU when not publishing. In this screenshot, the subscriber is consuming 0 messages and doesn't publish anything at all, but uses 18% CPU:

<img width="253" alt="Screenshot of subscriber-only process using 18.5% CPU while consuming no messages" src="https://user-images.githubusercontent.com/108205/78373667-e6730d80-7598-11ea-8d3b-273658e75ad7.png">

It does increase `publish` throughput by about 3% in my benchmark using 128-byte messages, but it doesn't achieve your goal of using less CPU at idle. It might be the difference between Rust using threads vs Crystal using fibers for sleeping such small durations because a _lot_ of the CPU time is spent in managing the event loop:

<img width="688" alt="Screenshot of a sampling profiler on the idle subscriber process" src="https://user-images.githubusercontent.com/108205/78375643-1de2b980-759b-11ea-9079-8432297ea17f.png">

Reply:
That flush loop should only wake up when data is being published, hence the channel use as it was before. We can't do a spin loop. Rust one also waits on a channel, so if nothing being published there is no CPU being used at all.

Reply:
We need a MPSC channel or we might have to dip below and use a conditional.

Reply:
I see what you meant earlier about the poor man's spin lock now. I didn't understand what the `thread::park` call in the Rust code was for, but now that makes a lot more sense.

Ended up adding back the channel because it's the only abstraction I could find that achieved that goal, but I used the `waiting_count` instead of checking whether the channel's queue was empty.

I also tried to use a `select` action:

```crystal
select
when @flush.send nil
else # When sending would block
end
```

… but this was a lot slower than the `if @waiting_count == 0` check. The `select` action only did about 3M 128-byte messages per second on my machine, but the simple `if` does 9.2M.

And just as a sanity check, my subscribe benchmark stays at 0% CPU while no messages are coming in.

Reply:
Surprisingly, my benchmarks with this implementation are faster than the Rust client, which did 7.8M/sec at 128 bytes per message.

```
publish/128             time:   [126.24 ns 128.16 ns 130.84 ns]
                        thrpt:  [7.6428 Melem/s 7.8027 Melem/s 7.9214 Melem/s]
```

The Go client matched the Crystal client in throughput at 9.2M, but it used 160% CPU to do it whereas the Crystal client only used 99%:

```
➜  nats.go git:(master) time go run examples/nats-bench/main.go -ms 128 -n 90000000 foo
Starting benchmark [msgs=90000000, msgsize=128, pubs=1, subs=0]
Pub stats: 9,244,356 msgs/sec ~ 1.10 GB/sec
go run examples/nats-bench/main.go -ms 128 -n 90000000 foo  11.78s user 4.83s system 160% cpu 10.326 total
```

Here are the results from the Crystal implementation:

```
➜  nats.cr git:(crystal-0.32-compatibility) ✗ time crystal run --release bin/bench_pub.cr -- 128 10
9,647,892 publishes/sec
crystal run --release bin/bench_pub.cr -- 128 10  9.42s user 1.62s system 98% cpu 11.210 total
```

Reply:
For the rust client you using release mode when running?

Reply:
I assume so. I used `cargo bench`, which I assume runs in release mode. Wouldn't make sense to run benchmarks without it. :-)